tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,501 @@
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import connection, models, transaction
+from django.db.models import Count, Q, Value
+from django.db.models.fields.json import KeyTextTransform, KeyTransform
+from django.db.utils import DatabaseError, IntegrityError
+from django.test import SimpleTestCase, TestCase
+
+from .models import JSONModel, NullableJSONModel, OrderedJSONModel
+
+
+@skipIf(connection.vendor == 'oracle', 'Oracle does not support meta ordering.')
+class TestModelMetaOrdering(TestCase):
+    def test_ordering_by_json_field_value(self):
+        OrderedJSONModel.objects.create(value={'b': 2})
+        OrderedJSONModel.objects.create(value={'a': 3})
+        objects = OrderedJSONModel.objects.all()
+        self.assertEqual(objects[0].value, {'a': 3})
+        self.assertEqual(objects[1].value, {'b': 2})
+
+
+class TestDefaultValue(SimpleTestCase):
+    def _set_default(self, value):
+        field = JSONModel._meta.get_field('value')
+        field.default = value
+        return field.check()
+
+    def tearDown(self):
+        self._set_default(dict)
+        return super().tearDown()
+
+    def test_mutable_default_value(self):
+        mutable_values = [
+            ['foo', 'bar', 123], {'foo': 'bar'},
+        ]
+        for value in mutable_values:
+            with self.subTest(value=value):
+                errors = self._set_default(value)
+                self.assertEqual(len(errors), 1)
+                self.assertIn('default should not be a mutable object', errors[0].msg)
+
+    def test_valid_default_value(self):
+        def callable_obj():
+            return {'it': 'works'}
+
+        valid_values = [
+            None, True, 123, 123.45, 'foo bar', ('foo', 'bar', 123), callable_obj,
+        ]
+        for value in valid_values:
+            with self.subTest(value=value):
+                errors = self._set_default(value)
+                self.assertEqual(len(errors), 0)
+
+
+class TestValidation(TestCase):
+    def _set_encoder_decoder(self, encoder, decoder):
+        field = JSONModel._meta.get_field('value')
+        field.encoder, field.decoder = encoder, decoder
+        return field.check()
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.uuid_value = {'uuid': uuid.UUID('{12345678-1234-5678-1234-567812345678}')}
+
+    def tearDown(self):
+        self._set_encoder_decoder(None, None)
+        return super().tearDown()
+
+    def test_validation_error(self):
+        field = models.JSONField()
+        with self.assertRaises(ValidationError) as err:
+            field.clean(self.uuid_value, None)
+        self.assertEqual(err.exception.code, 'invalid')
+        self.assertEqual(err.exception.message % err.exception.params, 'Value must be valid JSON.')
+
+    def test_not_serializable(self):
+        obj = JSONModel(value=self.uuid_value)
+        with transaction.atomic():
+            self.assertRaises(TypeError, obj.save)
```

Comment:
Shouldn't this be `with self.assertRaises`? Please also check if you can include the message.

Reply:
I wanted to emphasize we are asserting that an exception is raised, thus I put it into a statement rather than a context manager. Using a context manager is more flexible, though, since the code that raises the exception can be a whole block instead of just a method with no arguments.

Changed.